File: ffi/android/WormDB.kt
package dev.wormdb

/**
 * Embedded WormDB key/value store for Android — single-node, in-process, backed
 * by the Zig engine (WAL + snapshot persistence, std.crypto, no libsodium).
 *
 * Usage:
 *   val db = WormDB(context.filesDir.resolve("wormdb").absolutePath)
 *   db.set("greeting".toByteArray(), "hello".toByteArray())
 *   val v = db.get("greeting".toByteArray())   // ByteArray? (null if absent)
 *   db.use { ... }                             // AutoCloseable
 */
class WormDB(path: String, persistence: Int = PERSIST_FULL) : AutoCloseable {

    private val handle: Long = nativeOpen(path, persistence)

    init {
        require(handle != 0L) { "wormdb_open failed for path: $path" }
    }

    fun set(key: ByteArray, value: ByteArray, worm: Boolean = false) {
        check(nativeSet(handle, key, value, worm) == 0) { "wormdb_set failed" }
    }

    fun set(key: String, value: String, worm: Boolean = false) =
        set(key.toByteArray(), value.toByteArray(), worm)

    /** Returns the value bytes, or null if the key is absent. */
    fun get(key: ByteArray): ByteArray? = nativeGet(handle, key)

    fun getString(key: String): String? = get(key.toByteArray())?.toString(Charsets.UTF_8)

    fun delete(key: ByteArray) {
        check(nativeDelete(handle, key) == 0) { "wormdb_delete failed" }
    }

    fun delete(key: String) = delete(key.toByteArray())

    override fun close() = nativeClose(handle)

    private external fun nativeOpen(path: String, persistence: Int): Long
    private external fun nativeClose(handle: Long)
    private external fun nativeSet(handle: Long, key: ByteArray, value: ByteArray, worm: Boolean): Int
    private external fun nativeGet(handle: Long, key: ByteArray): ByteArray?
    private external fun nativeDelete(handle: Long, key: ByteArray): Int

    companion object {
        const val PERSIST_FULL = 0      // WAL every write + snapshots (durable)
        const val PERSIST_SNAPSHOT = 1  // load/save only
        const val PERSIST_MEMORY = 2    // in-memory only

        init {
            System.loadLibrary("wormdb_ffi") // the Zig engine
            System.loadLibrary("wormdb_jni") // this JNI bridge
        }
    }
}

File: ffi/swift/WormDB.swift
// Idiomatic Swift wrapper over the WormDB C FFI (ffi/wormdb.h).
//
// This is the kind of thin native layer an iOS app would embed: an in-process,
// single-node KV store backed by the Zig engine (WAL + snapshot persistence,
// std.crypto — no libsodium). No socket server, no clustering.

import Foundation

public final class WormDB {
    public enum Persistence: Int32 {
        case full = 0      // WAL on every write + snapshots (durable)
        case snapshot = 1  // load/save only, no per-write IO
        case memory = 2    // pure in-memory
    }

    public enum WormDBError: Error {
        case openFailed
        case writeFailed      // includes WORM violations
        case deleteFailed
    }

    private let handle: OpaquePointer

    /// Open or create a database rooted at `path` (an app-sandbox directory).
    public init(path: String, persistence: Persistence = .full) throws {
        guard let h = wormdb_open(path, persistence.rawValue) else {
            throw WormDBError.openFailed
        }
        handle = h
    }

    deinit { wormdb_close(handle) }

    public static var version: String { String(cString: wormdb_version()) }

    public func set(_ key: String, _ value: Data, worm: Bool = false) throws {
        let key = Array(key.utf8)
        let rc = key.withUnsafeBufferPointer { kp in
            value.withUnsafeBytes { (vp: UnsafeRawBufferPointer) -> Int32 in
                let vb = vp.bindMemory(to: UInt8.self)
                let fn = worm ? wormdb_set_worm : wormdb_set
                return fn(handle, kp.baseAddress, key.count, vb.baseAddress, value.count)
            }
        }
        if rc != 0 { throw WormDBError.writeFailed }
    }

    /// Convenience for string values.
    public func set(_ key: String, _ value: String, worm: Bool = false) throws {
        try set(key, Data(value.utf8), worm: worm)
    }

    public func get(_ key: String) throws -> Data? {
        let key = Array(key.utf8)
        var outPtr: UnsafeMutablePointer<UInt8>? = nil
        var outLen: Int = 0
        let rc = key.withUnsafeBufferPointer { kp in
            wormdb_get(handle, kp.baseAddress, key.count, &outPtr, &outLen)
        }
        if rc == 1 { return nil }                 // WORMDB_NOT_FOUND
        if rc != 0 { throw WormDBError.writeFailed }
        guard let p = outPtr else { return nil }
        defer { wormdb_free(p, outLen) }          // buffer is library-owned
        return Data(bytes: p, count: outLen)
    }

    public func getString(_ key: String) throws -> String? {
        guard let d = try get(key) else { return nil }
        return String(decoding: d, as: UTF8.self)
    }

    public func delete(_ key: String) throws {
        let key = Array(key.utf8)
        let rc = key.withUnsafeBufferPointer { kp in
            wormdb_delete(handle, kp.baseAddress, key.count)
        }
        if rc != 0 { throw WormDBError.deleteFailed }
    }
}
